### PR TITLE
Memory tweaks

### PR DESCRIPTION
**Pull request by mramato**
@shunter you might have some thoughts on this.  I'd like this in 1.14 unless there's a good reason to hold off.

1. Have `AssociativeArray.remove` actually delete the key from the object hash instead of just setting it to undefined, avoiding unbounded object growth.

2. If Entity geometry was removed and a geometry of the same type was never re-added back, the destroyed Primitive instance would remain in a closure, causing any JavaScript objects referenced by the Primitive to hang around, in the case of entities, this was the array of instance ids, which could be fairly large.

3. In order to reduce the change of 2 being a problem in other cases, I marked additional properties as undefined in `Primitive.destroy` to avoid objects hanging around after it is called even if the primitive itself is still inside of a closure.

I originally experimented with the idea of modifying `destroyObject` to more thoroughly clean up properties by setting them to undefined, but our use of ES5 properties makes this impossible.

Originally discussed [on the forum](https://groups.google.com/d/msg/cesium-dev/QkTpzAHZ5bA/R_g8fC75BQAJ).

Here's some code to test and verify the memory is being released.  Add and remove the entities and then take a Heap snapshot and verify that no Primitive or Entity objects are in the heap.  When you run this in master, there will be 1000 Entities and a Primitive stuck there.

```javascript
var viewer = new Cesium.Viewer('cesiumContainer');
var testDataSource = new Cesium.CustomDataSource('test');
viewer.dataSources.add(testDataSource);

var createQuantity = 1000;
var created = 0;
var deleted = 0;

// Add entities
function addEntities(){
    for (var i = created; i < (created + createQuantity); i++) {
        var entity = testDataSource.entities.getById(i);
        var lng = -80.12;
        var lat = 25.46;
        var name = 'test' + i;
        var type = 'type' + i;
        var description = ' test description words ' + i +
                          ' test description words ' + i +
                          ' test description words ' + i +
                          ' test description words ' + i +
                          ' test description words ' + i;

        // If no heat dot found, create it
        if (typeof entity === 'undefined') {
            entity = testDataSource.entities.add({
                id: i,
                name: name,
                entityType: type,
                description: description,
                rectangle: {
                    coordinates: Cesium.Rectangle.fromDegrees(lng - 0.0001, lat - 0.0001,
                                                              lng + 0.0001, lat + 0.0001),
                    material: Cesium.Color.WHITE.withAlpha(0.5),
                    extrudedHeight: 100
                }
            });
        }
    }
    created = created + createQuantity;
    viewer.zoomTo(testDataSource);
}

// Remove all
function removeAllEntities(){
    for (var i = 0; i < testDataSource.entities.values.length; i++) {
        var entity = testDataSource.entities.values[i];
        testDataSource.entities.remove(entity);
        deleted += 1;
        i--;
    }
}

Sandcastle.addToolbarButton('Add ' + createQuantity + ' Entities', function() {
    addEntities();
});

Sandcastle.addToolbarButton('Remove All Entities', function() {
    removeAllEntities();
});
```

### CODE REVIEW

File: Source/Scene/Primitive.js [hunk 1]
```diff
@@ -1500,6 +1500,12 @@ define([
         }
         this._pickIds = undefined;
 
+        this._instanceIds = undefined;
```
Comment (pjcozzi) [inline Source/Scene/Primitive.js]: Can you add a comment about why we're doing this here?

Comment (mramato) [inline Source/Scene/Primitive.js]: Done, I also updated changes and merged in master.

**Comment by mramato** [comment]
Bump to try and get this into 1.14 later today.

File: Source/Scene/Primitive.js [hunk 2]
```diff
@@ -1500,6 +1500,15 @@ define([
...
         this._pickIds = undefined;
 
+        //These objects may be fairly large and reference other large objects (like Entities)
+        //We explicitly set them to undefined here so that the memory can be freed
+        //even if a reference to the destroyed Primitive has been kept around.
+        this._instanceIds = undefined;
+        this._perInstanceAttributeCache = undefined;
+        this._perInstanceAttributeLocations = undefined;
+        this._attributeLocations = undefined;
+        this._dirtyAttributes = undefined;
```
Comment (emackey) [inline Source/Scene/Primitive.js]: I wonder if `destroyObject` should wipe out all keys in any destroyed object.  Performance hit but potential GC gain?

Comment (mramato) [inline Source/Scene/Primitive.js]: That was actually my first approach, but I ran into some problems with that approach.  I think it's a good idea to revisit the issue later on.

**Pull request merged**